add guide for custom xcom backends

kentdanas opened the pull request:
@vparekh94 @jwitz would love your review and feedback on this one!

jwitz commented on guides/custom-xcom-backends.md:
````diff
@@ -0,0 +1,285 @@
+    ge_batch_kwargs_list_pass >> ge_batch_kwargs_pass >> ge_checkpoint_pass
+```
+
+Now if we run this DAG, we will see three XCom files in our S3 bucket, one for each operator. With this implementation, we didn't need to enable XCom pickling, can version and access our XCom data easily, and are not filling up the Airflow metadata database, making this a much more sustainable way of using this popular operator.
````
```suggestion
When we run this DAG, we will see three XCom files in our S3 bucket, one for each operator. With this implementation, we didn't need to enable XCom pickling, we can version and access our XCom data easily, and we are not filling up the Airflow metadata database, making this a much more sustainable way of using this popular operator.
```
I see the word "Sustainable" throughout this article, and it's not clear to me what it means in this context. Could "reliable" or "scalable" be a better choice here?

kentdanas replied:
In this context it means that you could keep using XComs in this way over time without any manual intervention (like to clean up your metadata database, which you would likely have to do with normal xcoms). I don't think "reliable" or "scalable" captures the same thought here, but maybe there's a better word?

kentdanas commented on guides/custom-xcom-backends.md:
```diff
@@ -0,0 +1,285 @@
+
+XCom backends provide much more flexibility than you would have using traditional XComs with Airflow's metadata database. The biggest practical differences are that when using a custom backend your XComs do not need to be Json serializable (because you can implement your own serialization/deserialization and store the data however you want), and the size of your XCom information is not limited by ...
+
+All of this can be great if you want to pass information between your tasks flexibly and in a sustainable manner. However, note that even though a custom XCom backend will allow you to pass more data between your tasks, Airflow is still not designed to be a processing framework. XComs were meant to pass *messages* between your tasks, e.g. metadata or *small* amounts of data, not large datasets. H ...
```
Agree, makes sense

kentdanas commented on guides/custom-xcom-backends.md:
```diff
@@ -0,0 +1,285 @@
+- Create a bucket. In S3, make sure to block public access, enable versioning, and give the bucket a unique name.
+- Create a policy for your bucket that will allow Airflow to access it. Ensure the policy allows read and write actions over the bucket as shown in the screenshot below.
+
+![XCom Backend Policy](https://assets2.astronomer.io/main/guides/xcom/xcom_backend_policy.png)
```
I had no idea how to do this, thank you!

kentdanas commented on guides/custom-xcom-backends.md:
```diff
@@ -0,0 +1,285 @@
+6. Add an Airflow connection to connect to your backend as shown in step 2 above. Note that if you are using a secrets backend with Astronomer, you can add the connection there.
+7. Test your new XCom backend by running an appropriate DAG!
+
+For this example, we have created serialization and deserialization methods that take the results of the `GreatExpectationsOperator`, convert them to JSON, and save to a file on S3. Our custom backend class looks like this:
```
Yes, they were meant to be different. The first deals with pandas dataframes generally, and this one is specifically to handle the great expectations use case. I wanted to show a general example, and then a more specific one, and include multiple serialization/deserialization methods to give people a sense of what you could do with them,

kentdanas commented on guides/custom-xcom-backends.md:
```diff
@@ -0,0 +1,285 @@
+
+In December 2020, the Apache Airflow project made a huge step forward with the release of [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0). One of the main new features in this release is the [TaskFlow API](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#taskflow-api), which was introduced to solve the challenge of explicitly passing messages between Airfl ...
+
+One of the features of the TaskFlow API that increases the flexibility of XComs is support for a [custom XCom backend](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html?highlight=xcom#custom-xcom-backend). This means that rather than store XComs in Airflow's metadata database by default, you can push and pull XComs to and from an external system such as S3, GCS, or HDFS. You can ...
```
Yeah I see what you mean, I'll take it out here

kentdanas commented on guides/custom-xcom-backends.md:
```diff
@@ -0,0 +1,285 @@
+
+### Example Use Case: Great Expectations
+
+As a more practical example, using [Great Expectations](https://greatexpectations.io/) with Airflow is an ideal use case for implementing a custom XCom backend. Great Expectations is an open source Python-based data validation framework, and thanks to the [Great Expectations Airflow provider](https://registry.astronomer.io/providers/great-expectations), it integrates seamlessly with Airflow.
```
Good call, added a note in the section above.

kentdanas commented on guides/custom-xcom-backends.md:
```diff
@@ -0,0 +1,285 @@
+
+The first step to setting up a custom XCom backend is to configure the backend you want to use. The steps below apply to setting up an S3 backend on AWS, but the process would look similar if you used another service such as Azure blob storage, GCS, etc.
+
+- Create a bucket. In S3, make sure to block public access, enable versioning, and give the bucket a unique name.
```
I put "your S3 account" - you could say "console", but this can also be done programmatically so kept it more general

kentdanas commented on guides/custom-xcom-backends.md:
```diff
@@ -0,0 +1,285 @@
+
+### 2. Configure Your Airflow Environment
+
+The next step is configuring your Airflow environment to ensure it can connect to your custom backend. There are many ways you could accomplish this, some of which are more secure and sustainable than others. The end goal here is to define the relevant default connection such that it connects to your custom backend; for this example with S3, this means defining the `AIRFLOW_CONN_AWS_DEFAULT` envi ...
```
Modified your suggestion a little bit because it doesn't have to be a 'default'; otherwise the update looks great.

kentdanas commented:
> Thanks for sending this my way @kentdanas, and apologies in advance for flooding it with comments! Requesting changes because there were a few places where I requested adding some additional context or considerations that I couldn't write myself.

@jwitz thanks for your review! I accepted or addressed most of your comments. On the two different examples for serialization/deserialization methods, I included both of these on purpose. I wanted to keep the middle section very general and not applied to any specific use case; then included the great expectations one as something more applied and put it in the Astro platform section just as a way to highlight a different example. But it seems like maybe this wasn't clear. I do want to still include both of them, but maybe there's another way of organizing things to ensure there's no confusion?

jwitz commented on guides/custom-xcom-backends.md:
```diff
@@ -0,0 +1,300 @@
+1. Configure your backend (S3, GCS, etc.) as described in Section 1 from the previous example.
+2. Open an existing Astronomer project, or initialize a new one using `astro dev init` (if you don't already have the Astronomer CLI, install it using the [CLI Quickstart](https://www.astronomer.io/docs/cloud/stable/develop/cli-quickstart) before completing these steps).
+3. Add the XCom backend Python file with serialization/deserialization methods as shown below to the `include/` directory of your Astronomer project.
+4. Add the `AIRFLOW__CORE_XCOM_BACKEND` environment variable to your Astronomer project. For this example, it should look like `AIRFLOW__CORE__XCOM_BACKEND=include.s3_xcom_backend.S3XComBackend`. There are a few ways to add an environment variable to your Astronomer deployment, which are detailed in Astronomer's [Environment Variables guide](https://www.astronomer.io/docs/cloud/stable/deploy/envi ...
```
```suggestion
4. Add the `AIRFLOW__CORE_XCOM_BACKEND` environment variable to your Astronomer project. For this example, it should look like `AIRFLOW__CORE__XCOM_BACKEND=include.s3_xcom_backend.S3XComBackend`. There are a few ways to add an environment variable to your Astronomer Deployment, which are detailed in Astronomer's [Environment Variables guide](https://www.astronomer.io/docs/cloud/stable/deploy/environment-variables). In this case, we added the variable to our Dockerfile.
```

jwitz reviewed (approved):
Looks awesome @kentdanas ! Just had one minor style thing to add, but I'm preemptively approving. As long as you make it clear in the steps where to pull the sample code from, I think it's fine to have two examples.
